Content.Server/GameObjects/EntitySystems/ItemCabinetSystem.cs

```diff
@@ -0,0 +1,188 @@
+            }
+            else
+            {
+                RaiseLocalEvent(uid, new TryInsertItemCabinetEvent(args.User, args.Used), false);
+            }
+
+            args.Handled = true;
+        }
+
+        private void OnInteractHand(EntityUid uid, ItemCabinetComponent comp, InteractHandEvent args)
+        {
+            args.Handled = true;
+            if (comp.Opened)
+            {
+                if (comp.ItemContainer.ContainedEntity == null)
+                {
+                    RaiseLocalEvent(uid, new ToggleItemCabinetEvent(), false);
+                    return;
+                }
+                RaiseLocalEvent(uid, new TryEjectItemCabinetEvent(args.User), false);
+            }
+            else
+            {
+                RaiseLocalEvent(uid, new ToggleItemCabinetEvent(), false);
+            }
+        }
+
+        private void OnActivateInWorld(EntityUid uid, ItemCabinetComponent comp, ActivateInWorldEvent args)
+        {
+            args.Handled = true;
+            RaiseLocalEvent(uid, new ToggleItemCabinetEvent(), false);
+        }
+
+        /// <summary>
+        ///     Toggles the ItemCabinet's state.
+        /// </summary>
+        private void OnToggleItemCabinet(EntityUid uid, ItemCabinetComponent comp, ToggleItemCabinetEvent args)
+        {
+            comp.Opened = !comp.Opened;
+            ClickLatchSound(comp);
+            UpdateVisuals(comp);
+        }
+
+        /// <summary>
+        ///     Tries to insert an entity into the ItemCabinet's slot from the user's hands.
+        /// </summary>
+        private static void OnTryInsertItemCabinet(EntityUid uid, ItemCabinetComponent comp, TryInsertItemCabinetEvent args)
+        {
+            if (comp.ItemContainer.ContainedEntity != null || args.Cancelled || (comp.Whitelist != null && !comp.Whitelist.IsValid(args.Item)))
+            {
+                return;
+            }
+            var handsComponent = args.User.GetComponent<HandsComponent>();
+
+            if (!handsComponent.Drop(args.Item, comp.ItemContainer))
+            {
+                return;
+            }
+
+            UpdateVisuals(comp);
+        }
+
+        /// <summary>
+        ///     Tries to eject the ItemCabinet's item, either into the user's hands or onto the floor.
+        /// </summary>
+        private static void OnTryEjectItemCabinet(EntityUid uid, ItemCabinetComponent comp, TryEjectItemCabinetEvent args)
+        {
+            if (comp.ItemContainer.ContainedEntity == null || args.Cancelled)
+                return;
+            if (args.User.TryGetComponent(out HandsComponent? hands))
+            {
+                comp.Owner.PopupMessage(args.User,
+                    Loc.GetString("comp-item-cabinet-successfully-taken", ("item", comp.ItemContainer.ContainedEntity) , ("cabinet", comp.Owner)));
+                hands.PutInHandOrDrop(comp.ItemContainer.ContainedEntity.GetComponent<ItemComponent>());
```

Comment:
I don't think ItemCabinet accounts for ItemComponent existing except for here so it seems like it'd be easy to make this throw, e.g. if someone makes it spawn a prototype without it.

Reply:
i think the previous code just worked off of the assumption that no one would spawn something that isn't an item inside of an extinguisher cabinet, but i will add it anyway